connectivity/FEATURE_BLE/source/generic/GapImpl.cpp

```diff
@@ -2239,43 +2178,274 @@ void Gap::on_extended_advertising_report(
+    if (!_event_handler) {
         return;
     }
+
+    peer_address_type_t peer_address_type = address_type ? (peer_address_type_t::type) address_type->value()
+                                                         : peer_address_type_t::ANONYMOUS;
+
+    AdvertisingReportEvent event(
+        BLE_ERROR_NONE,
+        event_type,
+        peer_address_type,
+        address,
+        primary_phy,
+        secondary_phy ? *secondary_phy : phy_t::NONE,
+        advertising_sid,
+        tx_power,
+        rssi,
+        periodic_advertising_interval,
+        (ble::peer_address_type_t::type) direct_address_type.value(),
+        (address_t &) direct_address,
+        make_Span(data, data_length)
+    );
+
+    signal_advertising_report(
+        event
+    );
+}
+
+#if BLE_FEATURE_CONNECTABLE
+void Gap::signal_connection_complete(
+    ConnectionCompleteEvent& event
+)
+{
+#if BLE_FEATURE_PRIVACY
+    bool address_resolved = false;
+
+    /* if address resolution is not needed or already handled then the address is already resolved */
+    if (!_privacy_enabled ||
+        _address_registry.is_controller_privacy_supported() ||
+        !is_random_private_resolvable_address(event.getPeerAddress())) {
+        address_resolved = true;
+    } else {
+#if BLE_ROLE_CENTRAL
+        if (event.getOwnRole() == connection_role_t::CENTRAL) {
+            if (_central_privacy_configuration.resolution_strategy == central_privacy_configuration_t::DO_NOT_RESOLVE) {
+                address_resolved = true;
+            }
+        }
+#endif BLE_ROLE_CENTRAL
+
+#if BLE_ROLE_PERIPHERAL
+        if (event.getOwnRole() == connection_role_t::PERIPHERAL) {
+            if (_peripheral_privacy_configuration.resolution_strategy ==
+                peripheral_privacy_configuration_t::DO_NOT_RESOLVE) {
+                address_resolved = true;
+            }
+        }
+#endif // BLE_ROLE_PERIPHERAL
+    }
+
+    const address_t *peer_address;
+    target_peer_address_type_t peer_address_type(target_peer_address_type_t::RANDOM);
+
+    /* first try to resolve synchronously in cache */
+    if (!address_resolved) {
+        address_resolved = _address_registry.resolve_address_in_cache(
+            event.getPeerAddress(),
+            &peer_address_type,
+            &peer_address
+        );
+
+        if (address_resolved) {
+            conclude_signal_connection_complete_after_address_resolution(
+                event,
+                peer_address_type,
+                peer_address
+            );
+            return;
+        }
+    }
+
+    /* if successful then proceed to call the handler immediately same as for when privacy is disabled */
+    if (address_resolved) {
+        _event_handler->onConnectionComplete(
+            event
+        );
+    } else {
+        _address_registry.queue_resolve_address(event.getPeerAddress());
+
+        PendingConnectionCompleteEvent *pending_event = new(std::nothrow) PendingConnectionCompleteEvent(
```

Comment:
I'd prefer to have a single allocation for the list entry, the PendingConnectionCompleteEvent and its inner ConnectionCompleteEvent.